resolver/resolver.cc

```diff
@@ -1452,6 +1551,24 @@ class ResolveConstantsWalk {
             return;
         }
 
+        // Populate local first definitions table for out-of-order reference checking.
+        // This needs to happen *iff* the symbol is:
+        //   a). defined only in this file and nowhere else
+        //   b). defined multiple times in this file
+        //   c). this particular assign is the first occurrence of the symbol in the file
+        if (!ctx.file.data(ctx).isRBI() && loadScopeDepth_ == 0 &&
+            id->symbol.isOnlyDefinedInFile(ctx.state, ctx.file)) {
+            auto defLoc = id->symbol.loc(ctx).offsets();
+            auto declLoc = asgn.loc;
+
+            if (defLoc.beginPos() >= declLoc.endPos()) {
+                // When this condition is met, effectively it means the file has multiple definitions of the symbol
+                // and this is the first one. We need to use the first def for out-of-order checking, since any
+                // reference *before* the first def will be out of order.
+                firstDefinitionLocs->insert(std::make_pair(id->symbol, declLoc));
```

Comment:
Can we ENFORCE that there's nothing already there?

Reply:
Sure